src/raftstore/store/fsm/apply.rs

```diff
@@ -371,33 +376,60 @@ impl ApplyContext {
             self.write_to_db();
             self.prepare_for(delegate);
         }
-        self.wb_last_bytes = self.wb().data_size() as u64;
-        self.wb_last_keys = self.wb().count() as u64;
+        self.kv_wb_last_bytes = self.kv_wb().data_size() as u64;
+        self.kv_wb_last_keys = self.kv_wb().count() as u64;
     }
 
     /// Writes all the changes into RocksDB.
+    #[allow(clippy::useless_let_if_seq)]
     pub fn write_to_db(&mut self) {
-        if self.wb.as_ref().map_or(false, |wb| !wb.is_empty()) {
+        if self.kv_wb.as_ref().map_or(false, |wb| !wb.is_empty()) {
             let mut write_opts = WriteOptions::new();
-            write_opts.set_sync(self.enable_sync_log && self.sync_log_hint);
+            write_opts.set_sync(self.sync_log_hint);
             self.engines
                 .kv
-                .write_opt(self.wb(), &write_opts)
+                .write_opt(self.kv_wb(), &write_opts)
                 .unwrap_or_else(|e| {
                     panic!("failed to write to engine: {:?}", e);
                 });
-            self.sync_log_hint = false;
-            let data_size = self.wb().data_size();
+            let data_size = self.kv_wb().data_size();
             if data_size > APPLY_WB_SHRINK_SIZE {
                 // Control the memory usage for the WriteBatch.
-                self.wb = Some(WriteBatch::with_capacity(DEFAULT_APPLY_WB_SIZE));
+                self.kv_wb = Some(WriteBatch::with_capacity(DEFAULT_APPLY_WB_SIZE));
             } else {
                 // Clear data, reuse the WriteBatch, this can reduce memory allocations and deallocations.
-                self.wb().clear();
+                self.kv_wb().clear();
             }
-            self.wb_last_bytes = 0;
-            self.wb_last_keys = 0;
-        }
+            self.kv_wb_last_bytes = 0;
+            self.kv_wb_last_keys = 0;
+        }
+        if self.raft_wb.as_ref().map_or(false, |wb| !wb.is_empty()) {
+            if self.sync_log_hint {
+                let mut write_opts = WriteOptions::new();
+                write_opts.set_sync(true);
+                self.engines
+                    .raft
+                    .write_opt(self.raft_wb(), &write_opts)
+                    .unwrap_or_else(|e| {
+                        panic!("failed to write to engine: {:?}", e);
+                    });
+            }
+            // Delegates always write apply state when they apply raft cmds,
+            // so if there is no sync cmd for a long time, this may cause a
+            // large raft write batch which causes high write duration.
+            // To keep raft write batch small we can clear it if there is no
+            // sync cmd in the current cmd batch.
```

Comment:
I'm curious about that when the `sync_log_hint` is true, but we lost some apply states for some regions. If some of these regions are unlucky and always lost apply states, what will happen?

Reply:
It restarts apply from the latest saved applied index, since all commands are idempotent, it will be consistent with other peers eventually.